tasks/task_manager.cc

```diff
@@ -17,6 +17,73 @@ namespace tasks {
 
 logging::logger tmlogger("task_manager");
 
+bool task_manager::task::children::all_finished() const noexcept {
+    return _children.empty();
+}
+
+size_t task_manager::task::children::size() const noexcept {
+    return _children.size() + _finished_children.size();
+}
+
+future<> task_manager::task::children::add_child(foreign_task_ptr task) {
+    rwlock::holder exclusive_holder = co_await _lock.hold_write_lock();
+
+    auto id = task->id();
+    _children.emplace(id, std::move(task));
+}
+
+future<> task_manager::task::children::mark_as_finished(task_id id) const {
+    rwlock::holder exclusive_holder = co_await _lock.hold_write_lock();
+
+    auto& child_ptr = _children.at(id);
```

Comment:
Since this function holds the write lock it can rely on stable iterators the _children map,
so you can use find() to lookup the iterator to the child and reuse it later on to erase it from _children.
The benefit is that according to https://en.cppreference.com/w/cpp/container/unordered_map/erase erasing using the iterator throws nothing (although it is not marked as noexcept, but we can put it in a noexcept block just in case).